man/man1/runbody1.1
.TH RUNBODY1 1NEMO "1 February 2025"

.SH "NAME"
runbody1, nbody1 \- (frontend to) Aarseth N-body code

.SH "SYNOPSIS"
.nf
\fBrunbody1\fP [parameter=value]
\fBnbody1  \fP\fI < nbody1.in\fP
.fi

.SH "DESCRIPTION"
\fBrunbody1\fP is a NEMO frontend to Aarseth's NBODY1 (fortran) program,
running the code in a run directory where all files are stored. See
also \fInbody1(1NEMO)\fP.
.PP
The new version 2.x interface tries to mimic to run NBODY1 in virial units
with a default input and output.
In previous versions the defaults were more geared towards
running the classic interface where data were scaled and times were measured
in crossing times, making comparisons with other codes in NEMO much harder.
This however means it is now more work to make NBODY1 run in classic mode.
Parameters names and meaning have not changed, except for the introduction
of kz(15), to work in code (virial) units.

.SH "PARAMETERS"
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBin=\fP
input file (optional, see \fBnbody=\fP to generate your own input 
data on the fly). If input data are given, you must also set KZ(4)=2 (see 
below).   WARNING:  The input snapshot will be mass-weight centered before
the integration started by NBODY1.
.br
Default: not used.
.TP
\fBoutdir=\fP
Output run directory. This directory should not exist, although
can be overriden with NEMO's standard \fBerror=\fP system keyword.
.br
No default.
.TP
\fBnbody=\fP
** Total particle number. The maximum value (NMAX) is compiled into
the NBODY1 program. Note that even if input files are supplied, this
parameter is needed.
.br
No default.
.TP
\fBnfix=\fP
Output frequency of data save (KZ(3)) or binaries (KZ(6)).
.br
[Default: 1]
.TP
\fBnrand=\fP
Random number sequence skip.
.br
[Default: 0]
.TP
\fBnrun=\fP
Run identification index. Not really used for anything, just a relic from
nameless Aarseth-type files. The \fBoutdir\fP now has this function.
.br
[Default: 1]
.TP
\fBeta=\fP
** Time-step parameter for force polynomial.
.br
[Default: 0.02]
.TP
\fBdeltat=\fP
Output time interval in units of the crossing time (M**2.5/(2E)**1.5).
.br
[Default: 0.25]
.TP
\fBtcrit=\fP
Termination time in units of the crossing time.
.br
[Default: 5]
.TP
\fBqe=\fP
Energy tolerance (restart if DE/E > 5*QE & KZ(2) > 1)
.br
[Default: 2e-5]
.TP
\fBeps=\fP
Potential softening parameter     
.br
[Default: 0.01]
.TP
\fBkz=\fP
** Non-zero options for alternative paths (see below) 
[Default: 1,2,1,2,0,0,0,0,0,1,1,0,0,0,1]
.nf
.ta +0.5i
   1  COMMON save on unit 1 if TCOMP > CPU or if TIME > TCRIT. [1]
   2  COMMON save on unit 2 at output (=1); restart if DE/E > 5*QE (=2). [2]
   3  Basic data written to unit 3 at output time (frequency NFIX). [1]
   4  Initial conditions on unit 4 (=1: output; =2: input). [2]
   5  Initial conditions (=0: uniform & isotropic; =1: Plummer). [0]
   6  Output of significant binaries. [0]
   7  Output of movie frames on unit 7. [0]
   8  Generation of two subsystems (merger experiment). [0]
   9  Individual bodies printed at output time (MIN(5**KZ9,N)). [0]
  10  No scaling of initial conditions. [1]
  11  Modification of ETA by tolerance QE. [1]
  12  Initial parameters for binary orbit. [0]
  13  Escaper removal (R > 2*RTIDE; RTIDE = 10*RSCALE). [0]
  14  Adjustment of coordinates & velocities to c.m. condition. [0]
  15  Use code units for tcrit/deltat [1]
.fi
.TP
\fBalphas=\fP
** Power-law index for initial mass function (routine DATA)
.br
[Default: 2.3]
.TP
\fBbody1=\fP
Maximum particle mass before scaling   
.br
[Default: 5.0]
.TP
\fBbodyn=\fP
Minimum particle mass before scaling   
.br
[Default: 1.0]
.TP
\fBq=\fP
** Virial ratio (q=0.5 for virial equilibrium)  
.br
[Default: 0]
.TP
\fBvxrot=\fP
XY-velocity scaling factor (> 0 for solid-body rotation)
.br
[Default: 0]
.TP
\fBvzrot=\fP
Z-velocity scaling factor (not used if VXROT =
.br
[Default: 0]
.TP
\fBrbar=\fP
Virial radius in pc (for scaling to physical
.br
[Default: 1]
.TP
\fBzmbar=\fP
Mean mass in solar units   
.br
[Default: 1]
.TP
\fBxcm=\fP
** Displacement for subsystem (routine SCALE; KZ(8)) 
.TP
\fBecc=\fP
Eccentricity of relative motion for subsystem (ECC =< 1)
.TP
\fBkstart=\fP
** Run mode: 1=new >1 restart, with increasing options of resetting 
variables. Only new runs are currently supported.
.TP
\fBtcomp=\fP
not supposed to be used, but counts max computing time alotted.
.TP
\fBKZ#=\fP
NEMO indexed keyword (see \fIgetparam(3NEMO)\fP) to override \fBkz=\fP, for example KZ4=1, while keeping all other \fBkz=\fP's.
.br
Default: not used
.TP
\fBexe=\fP
Name of the (fortran) executable to run.
.br
[Default: nbody1]

.SH "BUGS"
Scaling can cause output to become out of bounds.
.PP
Coding the the input-order logic in the frontend program can be hazardous
to your health. It will depend on how the NBODY1 main program is structured.

.SH "INPUT"
Order of input lines in "nbody1.in" for a new run (KSTART=1)
.nf

          variables                   condition             where
  ----------------------------    -------------------	  ---------
  KSTART TCOMP                                            MAIN
  Nbody NFIX NRAND NRUN                                   INPUT
  ETA DELTAT TCRIT QE EPS                                 INPUT
  KZ(1..15)                                               INPUT
      ALPHAS BODY1 BODYN          KZ(4).NE.2              DATA
      SEMI ECC                    KZ(12).NE.0             DATA
  Q VXROT VZROT RBAR ZMBAR                                SCALE
      NFRAME DELTAF               KZ(7).GT.0              SCALE
      XCM ECC                     KZ(8).GT.0              SUBSYS
.fi

.SH "CAVEATS"
The input snapshot is (mass-weighted) centered before integration is started.
Not clear why kz(14) still gives some different answers.
.PP
By decreasing TCRIT, you \fIobserve\fP the integrations at finer details, thus
re-syncing all particles again. This will have an effect on orbits

.SH "EXAMPLES"
If you want to use runnbody1 as a NEMO tool with snapshots in and out, here
is an example with a 128-body plummer sphere:
.nf
   mkplummer p128 128
   runbody1 p128 run1 deltat=0.1 tcrit=10
.fi
This will create a run1/OUT3.snap. Code units are now used, not crossing time units.

.PP
If however you want to use the more original NBODY1 flavored run, pay attention to the kz= vector
as the default will not suffice anymore.
.nf
   runbody1 outdir=run2 nbody=128 deltat=0.1 tcrit=10 kz=1,2,1,1,1
.fi
Note that requesting tcript=10 you will see the final snapshot written at time=28.2843
because the crossing time is about 2.828 and tcrit is measured in terms of crossing times,
as opposed to code times for the default run with kz(15)=1.

.SH "BENCHMARK"
The command
.nf
     cd $NEMO/src/nbody/evolve/aarseth/nbody1/
     make bench
.fi
will run a standard 1280 body plummer sphere benchark to T=10, typically taking 5-10 seconds.

.SH "CAVEATS"
\fBnbody1\fP still uses the old-style individual time-steps with force polynomials, not the
\fIimproved\fP hermite integrater that is embedded in later versions of the NBODYx series. However,
Sverre has made versions available with a hermite integrator, see for example
the micky-mouse version \fInbody04h\fP described in \fInbody0(1NEMO)\fP.

.SH "SEE ALSO"
nbody0(1NEMO), runbody2(1NEMO), runbody4(1NEMO), runbody6(1NEMO), snapshot(5NEMO), u3tos(1NEMO), stou4(1NEMO)

.SH "FILES"
.nf
.ta +3i
$NEMO/src/nbody/evolve/aarseth/nbody1	nbody code
$NEMO/src/nbody/evolve/aarseth/tools	various tools code
.ta +2i
$outdir/fort.1                       	restart dump (compile time dep. size)
$outdir/fort.2                         	restart dump (compile time dep. size)
$outdir/fort.4                         	restart dump (compile time dep. size) (see \fIstou4(1NEMO)\fP and \fIu4tos(1NEMO)\fP)
$outdir/fort.7                         	formatted frame files (option KZ(7))
$outdir/OUT3                        	particle dump (see \fIu3tos(1NEMO)\fP)
.fi

.SH "AUTHOR"
Peter Teuben

.SH "HISTORY"
.nf
.ta +1.25i +4.5i
4-mar-98	V1.1 documented, and changed out= to outdir=, n= to nbody= PJT
17-mar-04	V1.1a fixed bug when in= was specified
17-mar-06	V1.2 using fullname for in=
7-feb-2019	V2.0 better defaults for snapshot input	PJT
14-feb-2019	V2.1 default run now computes nbody= and converts to snapshot	PJT
20-jan-2024	formatting	PJT
.fi


src/nbody/evolve/aarseth/tools/runbody1.c
/*
 *  MAIN.C: NEMO driver for nbody1
 *	
 *  4-mar-98    cloned off runbody2, added KZ_ parameters     PJT
 *  17-mar-04   using in= didn't enable kz(4) properly        PJT
 *  17-mar-06   using fullname()                              PJT
 *  17-sep-2013  V1.3   using new run interface               PJT
 *   9-feb-2017  V2.0   better defaults for snapshot input    PJT
 *
 *
 */

#include <stdinc.h>
#include <getparam.h>
#include <history.h>
#include <filefn.h>
#include <run.h>

#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>

#include <vectmath.h>
#include <filestruct.h>

#include <snapshot/snapshot.h>
#include <snapshot/body.h>
#include <snapshot/get_snap.c>


string defv[] = {
    "in=\n            input file (optional - see nbody= ) ",
    "outdir=???\n     output run directory (required)",

    "nbody=\n         Total particle number (<= NMAX) if in= not given",
    "nfix=1\n         Output frequency of data save or binaries; KZ(3 & 6)",
    "nrand=0\n        Random number sequence skip",
    "nrun=1\n         Run identification index",

    "eta=0.02\n       Time-step parameter for irregular force polynomial",
    "deltat=0.25\n    Output time interval in units of the crossing time",
    "tcrit=2.0\n      Termination time in units of the crossing time",
    "qe=0.00002\n     Energy tolerance (restart if DE/E > 5*QE & KZ(2) > 1)",
    "eps=0.05\n       Potential softening parameter",


    "kz=1,2,1,2,0,1,0,0,0,1,1,0,0,0,1\n  Non-zero options for alternative paths (see below)\n"
      "   1  COMMON save on unit 1 if TCOMP > CPU or if TIME > TCRIT.\n"
      "   2  COMMON save on unit 2 at output (=1); restart if DE/E > 5*QE (=2).\n"
      "   3  Basic data written to unit 3 at output time (frequency NFIX).\n"
      "   4  Initial conditions on unit 4 (=1: output; =2: input).\n"
      "   5  Initial conditions (=0: uniform & isotropic; =1: Plummer).\n"
      "   6  Output of significant binaries.\n"
      "   7  Output of movie frames on unit 7.\n"
      "   8  Generation of two subsystems (merger experiment).\n"
      "   9  Individual bodies printed at output time (MIN(5**KZ9,N)).\n"
      "  10  No scaling of initial conditions.\n"
      "  11  Modification of ETA by tolerance QE.\n"
      "  12  Initial parameters for binary orbit.\n"
      "  13  Escaper removal (R > 2*RTIDE; RTIDE = 10*RSCALE).\n"
      "  14  Adjustment of coordinates & velocities to c.m. condition.\n"
      "  15  Use code units for tcrit/deltat",
    
    "alphas=2.3\n     Power-law index for initial mass function",
    "body1=5.0\n      Maximum particle mass before scaling",
    "bodyn=1.0\n      Minimum particle mass before scaling",

    "q=0.0\n          Virial ratio (q=0.5 for virial equilibrium)",
    "vxrot=0.0\n      XY-velocity scaling factor (> 0 for solid-body rotation)",
    "vzrot=0.0\n      Z-velocity scaling factor (not used if VXROT = 0)",
    "rbar=1.0\n       Virial radius in pc (for scaling to physical units)",
    "zmbar=1.0\n      Mean mass in solar units",

    "xcm=\n           Displacement for subsystem (routine SCALE; KZ 17)",
    "ecc=\n           Eccentricity of relative motion for subsystem (ECC =< 1)",

    "kstart=1\n       Running mode (1=new 2=restart 3,4,5=restart w/ new par",
    "tcomp=40.0\n     Maximum allowed running time (minutes)",

    "KZ#=\n           [indexed] Override some kz= keywords",
    "exe=nbody1\n     Name of the executable",
    
    "VERSION=2.2a\n   27-may-2021 PJT",
    NULL,
};

string usage="front end for Aarseth nbody1 N-body code";

#define KZ_MAX  15


#define KZ_OUT   3
#define KZ_INI   4
#define KZ_MOV   7
#define KZ_MER   8
#define KZ_BIN  12
#define KZ_COM  14

void nemo_main(void)
{
    int nbody, nfix, nrand, nrun, kstart;
    real eta, deltat, tcrit, qe, eps, tcomp;
    int k, nkz, kz[KZ_MAX];
    real alphas, body1, bodyn;
    real q, vxrot, vzrot, rbar, zmbar;
    string exefile = getparam("exe");
    string parfile = "nbody1.in";
    string rundir = getparam("outdir");
    string infile, fname;
    char dname[256], runcmd[256];
    stream datstr, histr;

    kstart = getiparam("kstart");
    tcomp =  getdparam("tcomp");

    nfix = getiparam("nfix");
    nrand = getiparam("nrand");
    nrun = getiparam("nrun");

    eta = getdparam("eta");
    deltat = getdparam("deltat");
    tcrit = getdparam("tcrit");
    qe = getdparam("qe");
    eps = getdparam("eps");

    nkz = nemoinpi(getparam("kz"),kz,KZ_MAX);
    if (nkz < 0) error("%d Syntax error kz=%s",nkz,getparam("kz"));
    for (k=nkz; k<KZ_MAX; k++) kz[k]=0;
    for (k=0; k<KZ_MAX; k++) dprintf(1,"%d ",kz[k]);
    dprintf(1,"\n");

    for (k=0; k<KZ_MAX; k++) {
      if (indexparam("KZ",k+1)) {
	dprintf(0,"KZ %d=%d\n",k+1,getiparam_idx("KZ",k+1));
	kz[k] = getiparam_idx("KZ",k+1);
      }
    }
    

    alphas = getdparam("alphas");
    body1 = getdparam("body1");
    bodyn = getdparam("bodyn");

    q = getdparam("q");
    vxrot = getdparam("vxrot");
    vzrot = getdparam("vzrot");
    rbar = getdparam("rbar");
    zmbar = getdparam("zmbar");

    infile = getparam("in");
    fname = fullname(infile);

    run_mkdir(rundir);

    sprintf(dname,"%s/%s",rundir,parfile);
    datstr = stropen(dname,"w");

    if (hasvalue("in")) {
      stream instr = stropen(getparam("in"),"r");
      nbody = get_snap_nbody(instr);
      strclose(instr);
      dprintf(0,"Grabbing nbody=%d\n",nbody);
    } else {
      if (hasvalue("nbody"))
	nbody = getiparam("nbody");
      else
	error("nbody= needs to be specified");
    }

    /*  New Run */

  if (kstart == 1) {

    fprintf(datstr,"%d %g\n",kstart,tcomp);
    fprintf(datstr,"%d %d %d %d\n",nbody,nfix,nrand,nrun);
    fprintf(datstr,"%g %g %g %g %g\n",eta,deltat,tcrit,qe,eps);

    if (hasvalue("in") && kz[KZ_INI-1] != 2) {
      warning("Using in= and setting kz(%d)=2",KZ_INI);
      kz[KZ_INI-1] = 2;
    }

    for (k=0; k<KZ_MAX; k++) fprintf(datstr,"%d ",kz[k]);
    fprintf(datstr,"\n");

    if (kz[KZ_INI-1] != 2) {
        fprintf(datstr,"%g %g %g\n",alphas,body1,bodyn);
        if (hasvalue("in")) warning("Also gave in=, will be ignored");
    } else
    	if (!hasvalue("in")) warning("Did not supply in=");
    fprintf(datstr,"%g %g %g %g %g\n",q,vxrot,vzrot,rbar,zmbar);

    if (kz[KZ_BIN-1] > 0)
        fprintf(datstr,"%g %g\n",getdparam("xcm"), getdparam("ecc"));
    strclose(datstr);

    run_cd(rundir);
    histr = stropen("history","w");
    put_history(histr);
    strclose(histr);

    if (hasvalue("in")) {
	if (*infile == '-') {		/* do something special for pipes */
	  sprintf(runcmd,"stou4 %s",infile);
	} else {
	  sprintf(runcmd,"stou4 %s nbody=%d",fname,nbody);
	}
	dprintf(0,"%s\n",runcmd);
        if (system(runcmd)) error("Error converting input data");
	// @todo   grab nbody
    }

    sprintf(runcmd,"%s < %s",exefile,parfile);
    run_sh(runcmd);
    
    sprintf(runcmd,"u3tos in=OUT3 out=OUT3.snap mode=1");
    run_sh(runcmd);    
    
  } else {
    error("kstart=%d not yet supported",kstart);
  }
}

/*
 *	Order of input lines in "nbody1.in" for a new run (KSTART=1)
 *
 *          variables                   condition               where
 *  ----------------------------    -------------------
 *  KSTART TCOMP                                            nbody1.f    MAIN
 *  Nbody NFIX NRAND NRUN                                   input.f     INPUT
 *  ETA DELTAT TCRIT QE EPS                                 input.f     INPUT
 *  KZ(1..15)                                               input.f     INPUT
 *      ALPHAS BODY1 BODYN          KZ(4).NE.2              data.f      DATA
 *      SEMI ECC                    KZ(12).NE.0             data.f      DATA
 *  Q VXROT VZROT RBAR ZMBAR                                scale.f     SCALE
 *      NFRAME DELTAF               KZ(7).GT.0              scale.f     SCALE
 *      XCM ECC                     KZ(8).GT.0              subsys.f    SUBSYS
 *
 */
 
